ssh-pairing-server.c

```diff
@@ -59,13 +98,21 @@ int main(int argc, char *argv[])
 	// the callback-based pubkey handling is not possible:
 	// https://gitlab.com/libssh/libssh-mirror/-/issues/146
 
+	/*
+	 * I don't really get why the INTERACTIVE auth method is offered in
+	 * the first place. It looks like the only purpose is to send the
+	 * "Received %d public keys" message to the connecting client. This
+	 * should be documented better.
```

Comment:
Almost. Without offering keyboard-interactive, the client shows a `Permission denied` message. It still works, but it's misleading. This is explained in the readme as well.

Reply:
I see. A short comment in the code about this wouldn't hurt.

Also, at least in my mind (as an expert only, maybe), the message `Permission denied` makes complete sense, since the login is not supposed to succeed at this stage.

Reply:
But there still seems to be the problem what happens if the client does not accept keyboard-interactive, right? Then there will neither be the "improved error message" nor the verification string visible on the client side.

Reply:
Correct. In that case it's not distinguishable from the client whether the connection got accepted by ssh-pairing-server or a full sshd. In the README it's meanwhile documented that the message on the client is mandatory, so without keyboard-interactive this doesn't work (securely).

Reply:
In that case I'd say that the ssh-pairing-daemon should not record any keys if no keyboard-interactive reply was possible.

Reply:
Can do. Do I need to reintroduce the allocation + delayed print at the end again or is it fine to just `exit(1)` in that case while still printing the received keys? IMO it should be fine as the caller needs proper error handling in any case.

Reply:
The clean approach would be not to print out any keys until some level of trust and consistency is established.

Instead of concatenating one large string you could also use an array like `char* keys[MAX_KEY_COUNT]`. Sadly C programming is rather bad with managing dynamically growing data structures.

Reply:
That's true. But at this stage you _know_ that something is off, so why risk continue processing the data?

Reply:
It doesn't really process it though.

Reply:
All right, it _offers_ it for processing by another component. But why?

Reply:
Less complexity in the server component. I don't see any downside to offer it. Validation happens only after the process exits anyway.